feat: [DO NOT MERGE] support customer managed instance configurations

rajatbhatta opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-spanner/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

olavloite commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+    this((BuilderImpl) newBuilder(id).setDisplayName(displayName));
+  }
+
+  public static Builder newBuilder(InstanceConfigId id) {
```
I would suggest removing this (or otherwise making it package-private if we need it for internal reasons). BaseConfig is always required for custom InstanceConfigs.

rajatbhatta replied:
Will make this newBuilder private.

rajatbhatta replied:
This newBuilder is used to create InstanceConfigInfo object to be passed during updateInstanceConfig. We do not require a baseConfig in that case.

olavloite commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+    BuilderImpl(InstanceConfigInfo instanceConfigInfo) {
+      this.id = instanceConfigInfo.id;
+      this.displayName = instanceConfigInfo.displayName;
+      this.replicas = instanceConfigInfo.replicas;
```
Make a copy of the Lists / Maps etc.

rajatbhatta replied:
Done.

olavloite commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java:
```diff
@@ -875,6 +880,48 @@ private void acquireAdministrativeRequestsRateLimiter() {
+      String parent,
+      String instanceConfigId,
+      InstanceConfig instanceConfig,
+      @Nullable Boolean validateOnly)
```
Is there a difference between null and false in this case? If not, I would prefer using a boolean instead of a Boolean.

rajatbhatta replied:
If the user does not pass in any validateOnlyOption, this value will be null. Hence, we have @Nullable here. If we have this value is null, we need not to populate validateOnly field to the backend.

olavloite replied:
Makes sense to me.

olavloite commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerRpc.java:
```diff
@@ -34,10 +34,7 @@
-import com.google.spanner.admin.instance.v1.Instance;
-import com.google.spanner.admin.instance.v1.InstanceConfig;
-import com.google.spanner.admin.instance.v1.UpdateInstanceMetadata;
+import com.google.spanner.admin.instance.v1.*;
```
Google style requires us to never use wildcard imports: https://google.github.io/styleguide/javaguide.html

rajatbhatta replied:
Made this change.

olavloite commented on google-cloud-spanner/src/test/java/com/google/cloud/spanner/InstanceConfigTest.java:
```diff
@@ -63,26 +76,44 @@ public void testInstanceConfigFromProto() {
-                    .build()),
-            Arrays.asList("Leader Option 1", "Leader Option 2"),
-            client),
+            (InstanceConfig.Builder)
```
I don't understand why we need a cast here?

rajatbhatta replied:
All the setters exist in base InstanceConfigInfo.Builder/BuilderImpl class and return InstanceConfigInfo.Builder object. We just inherit these setters in InstanceConfig.Builder. When we use these setters, the object returned is of type InstanceConfigInfo.Builder and we cast it to more specific InstanceConfig.Builder.

olavloite replied:
Ok, that makes sense, but I think that shows a small (potential?) problem with the current implementation: Users should not be forced to do any casts when working with the builder. So if the implementation in the abstract base class (`InstanceConfigInfo`) returns the abstract class, but all our public methods require the use of the concrete subclass (`InstanceConfig`), then we should make sure that the methods that are visible to users als return an instance of `InstanceConfig.Builder` and not `InstanceConfigInfo.Builder`.
Could you take a look at `InstanceConfig` and `InstanceConfigInfo` how that is handled there, and modify this test case (and potentially the implementation) to show that users do not need to do any casting when working with the builder(s)?

rajatbhatta replied:
Thanks for the explanation! 

The solution I've implemented now to handle this is to override the setters in InstanceConfig.Builder class and make them return InstanceConfig.Builder.

For example, overridden setDisplayName in InstanceConfig.Builder would look something like this:
```
@Override
public Builder setDisplayName(String displayName) {
  super.setDisplayName(displayName);
  return this;
}
```

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+      // Custom configurations contain all the replicas of the base config and at least one optional
+      // replica.
+      this.replicas = this.baseConfig.getReplicas();
+      this.replicas.add(baseConfig.getOptionalReplicas().get(0));
```
Oops, had to remove this after adding the addReadOnlyReplicas support. Will remove this method.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+    private void addAllReplicasFromBaseConfig() {
+      // Custom configurations contain all the replicas of the base config and at least one optional
+      // replica.
+      this.replicas = this.baseConfig.getReplicas();
```
This method is not needed now, after adding support for addReadOnlyReplicas method. Will remove it.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+    }
+
+    @Override
+    public Builder addAllReplicas(List<ReplicaInfo> replicas) {
```
Agreed. Will change this to protected, as the user is setting replicas through newBuilder(InstanceConfigId id, InstanceConfigInfo baseConfig) and addReadOnlyReplicas method. We'll need this method only during conversion from com.google.spanner.admin.instance.v1.InstanceConfig object to com.google.cloud.spanner.InstanceConfig object.

rajatbhatta replied:
Done.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+    }
+
+    @Override
+    protected Builder setConfigType(Type configType) {
```
It has been marked as protected.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+    return new BuilderImpl(id, baseConfig);
+  }
+
+  public InstanceConfigInfo(
```
Won't removing this or making it private be a breaking change? Users may be using this constructor for creating objects currently.

thiagotnunes replied:
+1, we can not change the visibility here. We could mark it as deprecated though (with docs pointing to the new method)

rajatbhatta replied:
Marked this constructor as deprecated with docs to point to newBuilder method.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java:
```diff
@@ -875,6 +880,48 @@ private void acquireAdministrativeRequestsRateLimiter() {
+
+  @Override
+  public OperationFuture<InstanceConfig, UpdateInstanceConfigMetadata> updateInstanceConfig(
+      InstanceConfig instanceConfig, @Nullable Boolean validateOnly, FieldMask fieldMask)
```
If the user does not pass in validateOnlyOption, this value will be null. Hence, we have Nullable here. If we have this value is null, we need not to populate validateOnly field to the backend (we need not even send false here, as that's the default behaviour if we do not pass in validateOnly).

olavloite replied:
Makes sense to me

rajatbhatta commented on google-cloud-spanner/clirr-ignored-differences.xml:
```diff
@@ -55,6 +95,11 @@
     <className>com/google/cloud/spanner/DatabaseAdminClient</className>
     <method>com.google.api.gax.longrunning.OperationFuture copyBackup(com.google.cloud.spanner.BackupId, com.google.cloud.spanner.Backup)</method>
   </difference>
+  <difference>
```
Done.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/Options.java:
```diff
@@ -151,6 +170,14 @@ public static ListOption filter(String filter) {
     return new FilterOption(filter);
   }
 
+  public static DeleteAdminAPIOption etag(String etag) {
```
Added.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+  public abstract static class Builder {
+    public abstract Builder setDisplayName(String displayName);
+
+    public abstract Builder addAllReplicas(List<ReplicaInfo> replicas);
```
Changed to protected.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+
+    @Override
+    protected Builder addAllOptionalReplicas(List<ReplicaInfo> optionalReplicas) {
+      this.optionalReplicas = optionalReplicas;
```
Changed to `set`.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+
+    @Override
+    public Builder addAllLeaderOptions(List<String> leaderOptions) {
+      this.leaderOptions = leaderOptions;
```
Changed to `set`.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+    }
+  }
+
+  public InstanceConfigInfo(InstanceConfigId id, String displayName) {
```
Won't making it private create a breaking change? Users may be using this constructor for creating objects currently.

thiagotnunes replied:
+1, we can not change the visibility here. We could mark it as deprecated though (with docs pointing to the new method)

rajatbhatta replied:
Marked this constructor as deprecated with docs to point to newBuilder method.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+  InstanceConfigInfo(BuilderImpl builder) {
+    this.id = builder.id;
+    this.displayName = builder.displayName;
+    this.replicas = builder.replicas;
```
Done.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+    }
+
+    @Override
+    public Builder setBaseConfig(InstanceConfigInfo baseConfig) {
```
Changed it to protected.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+    }
+
+    @Override
+    public Builder addAllLeaderOptions(List<String> leaderOptions) {
```
Added.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+
+    protected abstract Builder addAllOptionalReplicas(List<ReplicaInfo> optionalReplicas);
+
+    public abstract Builder setBaseConfig(InstanceConfigInfo baseConfig);
```
Changed to protected.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+
+    public abstract Builder setBaseConfig(InstanceConfigInfo baseConfig);
+
+    public abstract Builder addAllLeaderOptions(List<String> leaderOptions);
```
leader_options is not a read-only field. So, having it makes sense. Let me know if you think otherwise.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+
+    public abstract Builder putAllLabels(Map<String, String> labels);
+
+    public abstract Builder addReadOnlyReplicas(List<ReplicaInfo> readOnlyReplicas);
```
Done.

thiagotnunes commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceAdminClient.java:
```diff
@@ -19,14 +19,140 @@
+   *     instanceAdminClient.updateInstanceConfig(
+   *         instanceConfigInfo, ImmutableList.of(InstanceConfigField.DISPLAY_NAME));
+   *
+   * InstanceConfig instanceConfig = operation.get(5, TimeUnit.MINUTES);
```
nit: should we use the same timeout in minutes for all operations? This one uses 5 minutes, while above we use 10 minutes. Does it mean this one is expected to be faster? If not, I would argue to use the same timeout.

rajatbhatta replied:
Changing both the timeouts to 5 mins makes sense.

rajatbhatta replied:
Done.

thiagotnunes commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/Options.java:
```diff
@@ -51,6 +51,22 @@
+   * Marker interface to mark options applicable to Create, Update and Delete operations in admin
+   * API.
+   */
+  public interface CreateUpdateDeleteAdminAPIOption
```
Please use camel case for `Api` as well, per https://engdoc.corp.google.com/eng/doc/devguide/java/style/index.md?cl=head#s5.3-camel-case

Please change everywhere else too

rajatbhatta replied:
Done.

thiagotnunes commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceAdminClientImpl.java:
```diff
@@ -60,13 +65,81 @@ protected Policy fromPb(com.google.iam.v1.Policy policyPb) {
+  @Override
+  public OperationFuture<InstanceConfig, CreateInstanceConfigMetadata> createInstanceConfig(
+      InstanceConfigInfo instanceConfig, CreateAdminAPIOption... options) throws SpannerException {
+    final Options createAdminAPIOptions = Options.fromAdminAPIOptions(options);
```
Please use camel case `Api`, per https://engdoc.corp.google.com/eng/doc/devguide/java/style/index.md?cl=head#s5.3-camel-case

rajatbhatta replied:
Done.

thiagotnunes commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -81,17 +381,159 @@ public boolean equals(Object o) {
+    return builder.build();
+  }
+
+  static InstanceConfig.Type toProtoConfigType(Type type) {
```
Could this be private? If it is more open just for testing, could you add the `@VisibleForTesting` annotation?

rajatbhatta replied:
Making it private makes sense. It is not directly used in unit tests, so I think we need not add @VisibleForTesting annotation.

thiagotnunes commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -81,17 +381,159 @@ public boolean equals(Object o) {
+    }
+  }
+
+  static InstanceConfig.State toProtoState(State state) {
```
Could this be private? If it is more open just for testing, could you add the `@VisibleForTesting` annotation?

rajatbhatta replied:
Making it private makes sense. It is not directly used in unit tests, so I think we need not add @VisibleForTesting annotation.

thiagotnunes commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -81,17 +381,159 @@ public boolean equals(Object o) {
+    }
+  }
+
+  static com.google.cloud.spanner.InstanceConfig fromProto(
```
Could this be private? If it is more open just for testing, could you add the `@VisibleForTesting` annotation?

rajatbhatta replied:
This cannot be made private because the method is being used in InstanceAdminClientImpl.java in createInstanceConfig, updateInstanceConfig, etc.

thiagotnunes commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -81,17 +381,159 @@ public boolean equals(Object o) {
+    return builder.build();
+  }
+
+  static State fromProtoState(com.google.spanner.admin.instance.v1.InstanceConfig.State state) {
```
Could this be private? If it is more open just for testing, could you add the `@VisibleForTesting` annotation?

rajatbhatta replied:
Making it private makes sense. It is not directly used in unit tests, so I think we need not add @VisibleForTesting annotation.

thiagotnunes commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -81,17 +381,159 @@ public boolean equals(Object o) {
+    }
+  }
+
+  static Type fromProtoConfigType(com.google.spanner.admin.instance.v1.InstanceConfig.Type type) {
```
Could this be private? If it is more open just for testing, could you add the `@VisibleForTesting` annotation?

rajatbhatta replied:
Making it private makes sense. It is not directly used in unit tests, so I think we need not add `@VisibleForTesting` annotation.

thiagotnunes commented on pom.xml:
```diff
@@ -4,7 +4,7 @@
   <artifactId>google-cloud-spanner-parent</artifactId>
   <packaging>pom</packaging>
-  <version>6.23.2-SNAPSHOT</version><!-- {x-version-update:google-cloud-spanner:current} -->
+  <version>6.23.2-SNAPSHOT</version>
```
Why did we remove the metadata here?

rajatbhatta replied:
We need not remove this. Re-added the same.

thiagotnunes commented on google-cloud-spanner/clirr-ignored-differences.xml:
```diff
@@ -35,6 +35,46 @@
+  <difference>
+    <differenceType>7004</differenceType>
+    <className>com/google/cloud/spanner/InstanceConfigInfo</className>
+    <method>InstanceConfigInfo(com.google.cloud.spanner.InstanceConfigId, java.lang.String, java.util.List, java.util.List)</method>
```
I don't think we can have breaking changes in constructors, could you please overload it instead of changing it to avoid the breakage?

rajatbhatta replied:
This constructor has not been changed, and these clirr differences are not required here. Will remove these. Thanks for pointing this out.

rajatbhatta replied:
Removed these differences.

thiagotnunes commented on google-cloud-spanner/clirr-ignored-differences.xml:
```diff
@@ -35,6 +35,46 @@
+  <difference>
+    <differenceType>7004</differenceType>
+    <className>com/google/cloud/spanner/InstanceConfig</className>
+    <method>InstanceConfig(com.google.cloud.spanner.InstanceConfigId, java.lang.String, java.util.List, java.util.List, com.google.cloud.spanner.InstanceAdminClient)</method>
```
I don't think we can have breaking changes in constructors, could you please overload it instead of changing it to avoid the breakage?

rajatbhatta replied:
This constructor has not been changed, and these clirr differences are not required here. Will remove these. Thanks for pointing this out.

rajatbhatta replied:
Removed these differences.

rajatbhatta commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java:
```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+
+    BuilderImpl(InstanceConfigId id) {
+      this.id = id;
+      this.labels = new HashMap<>();
```
Thanks for a clear explanation. This is done.

thiagotnunes commented on google-cloud-spanner/src/main/java/com/google/cloud/spanner/Options.java:
```diff
@@ -151,6 +170,33 @@ public static ListOption filter(String filter) {
+   *   <li>{@link InstanceAdminClient#deleteInstanceConfig}
+   * </ul>
+   */
+  public static CreateUpdateDeleteAdminApiOption validateOnly(Boolean validateOnly) {
```
nit: should this be `boolean` or can we explicitly set a `null`?

thiagotnunes reviewed (approved):
LGTM

rajatbhatta commented:
Hi @thiagotnunes @olavloite : I've added support (both implementation and tests) for ListInstanceConfigOperations as well to this PR. Can you please review changes since previous code review? Thank you!

olavloite reviewed (approved):
The last changes also look good to me. I assume that we will add end-to-end tests either in a separate PR, or later in this PR once it is enabled on the backend?

rajatbhatta commented:
> The last changes also look good to me. I assume that we will add end-to-end tests either in a separate PR, or later in this PR once it is enabled on the backend?

Thanks for the review! Yes, we'll add end-to-end tests in a separate PR, along with the samples. We'll be writing integration tests using those samples (like in other clients).
